server/container_create_linux.go

```diff
@@ -702,95 +645,60 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrIface.Contai
 			Type:        "bind",
 			Source:      sb.HostnamePath(),
-			Destination: "/etc/hostname",
 			Options:     append(options, "bind"),
-		}
-		specgen.AddMount(mnt)
+		})
 	}
 
 	if !isInCRIMounts("/etc/hosts", containerConfig.GetMounts()) && hostNetwork(containerConfig) {
 		// Only bind mount for host netns and when CRI does not give us any hosts file
-		mnt = rspec.Mount{
+		ctr.SpecAddMount(rspec.Mount{
+			Destination: "/etc/hosts",
 			Type:        "bind",
 			Source:      "/etc/hosts",
-			Destination: "/etc/hosts",
 			Options:     append(options, "bind"),
-		}
-		specgen.AddMount(mnt)
+		})
 	}
 
 	if ctr.Privileged() {
-		setOCIBindMountsPrivileged(&specgen)
+		setOCIBindMountsPrivileged(specgen)
 	}
 
 	// Set hostname and add env for hostname
 	specgen.SetHostname(sb.Hostname())
 	specgen.AddProcessEnv("HOSTNAME", sb.Hostname())
 
-	specgen.AddAnnotation(annotations.Name, containerName)
-	specgen.AddAnnotation(annotations.ContainerID, containerID)
-	specgen.AddAnnotation(annotations.SandboxID, sb.ID())
-	specgen.AddAnnotation(annotations.SandboxName, sb.Name())
-	specgen.AddAnnotation(annotations.ContainerType, annotations.ContainerTypeContainer)
-	specgen.AddAnnotation(annotations.LogPath, logPath)
-	specgen.AddAnnotation(annotations.TTY, fmt.Sprintf("%v", containerConfig.Tty))
-	specgen.AddAnnotation(annotations.Stdin, fmt.Sprintf("%v", containerConfig.Stdin))
-	specgen.AddAnnotation(annotations.StdinOnce, fmt.Sprintf("%v", containerConfig.StdinOnce))
-	specgen.AddAnnotation(annotations.ResolvPath, sb.ResolvPath())
-	specgen.AddAnnotation(annotations.ContainerManager, lib.ContainerManagerCRIO)
-
 	created := time.Now()
-	specgen.AddAnnotation(annotations.Created, created.Format(time.RFC3339Nano))
-
-	metadataJSON, err := json.Marshal(metadata)
-	if err != nil {
-		return nil, err
-	}
-	specgen.AddAnnotation(annotations.Metadata, string(metadataJSON))
-
-	labelsJSON, err := json.Marshal(labels)
-	if err != nil {
-		return nil, err
-	}
-	specgen.AddAnnotation(annotations.Labels, string(labelsJSON))
-
-	kubeAnnotationsJSON, err := json.Marshal(kubeAnnotations)
-	if err != nil {
-		return nil, err
-	}
-	specgen.AddAnnotation(annotations.Annotations, string(kubeAnnotationsJSON))
-
-	spp := securityContext.GetSeccompProfilePath()
+	spp := containerConfig.GetLinux().GetSecurityContext().GetSeccompProfilePath()
 	if !ctr.Privileged() {
-		if err := s.setupSeccomp(ctx, &specgen, spp); err != nil {
+		if err := s.setupSeccomp(ctx, specgen, spp); err != nil {
 			return nil, err
 		}
 	}
-	specgen.AddAnnotation(annotations.SeccompProfilePath, spp)
 
 	mountPoint, err := s.StorageRuntimeServer().StartContainer(containerID)
 	if err != nil {
 		return nil, fmt.Errorf("failed to mount container %s(%s): %v", containerName, containerID, err)
 	}
+	err = ctr.SpecAddAnnotations(sb, containerVolumes, mountPoint, containerImageConfig.Config.StopSignal, imgResult, s.config.CgroupManager().IsSystemd(), node.SystemdHasCollectMode())
```

Comment:
we actually need  to do this after the defer below, otherwise, we risk not cleaning up the container in the storage. If this merges before you get to it, can you do a follow up?

Reply:
fixed !